src/lem/constrainer.rs

```diff
@@ -26,7 +27,7 @@ impl<F: LurkField> AllocationManager<F> {
     /// Checks if the allocation for a numeric variable has already been cached.
     /// If so, return the cached allocation variable. Allocate, cache and return
     /// otherwise.
-    pub(crate) fn alloc<CS: ConstraintSystem<F>>(
+    pub(crate) fn get_or_alloc_num<CS: ConstraintSystem<F>>(
```

Comment:
Hmmm, are these for constants? If so, I'm skeptical that this is going to be an efficient approach. If not, how is this conditional allocation a good idea? Probably I just don't fully get this yet.

Reply:
Yes, this is for constants such as the tags in a `MatchTag`. Why would it not be efficient?

Reply:
In the end, it can be efficient — the question is how much of the current implementation will remain, which might then become philosophical. Here are some things to watch out for:
- There should be only one allocation per circuit (not reduction circuit). I think that implies the `AllocationManager` needs to live longer than each reduction-circuit construction.
- lots of hash map lookups for constants

Those are both things the existing `GlobalAllocations` avoids. We can probably do 'better' than that too, but hopefully not 'worse'.